consensus/packets/tristatebitset.go

```diff
@@ -60,13 +63,151 @@ func (dbs *TriStateBitSet) ApplyChanges(changes []BitSetCell) {
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to get bitarray from cells")
+	}
+
+	tmpLen := array.Len()
+	totalSize := int(math.Round(float64((tmpLen*2)/8))+0.5) + 1 // first byte
+	var result *bytes.Buffer
+	firstByte = firstByte << 1
+	if bits.Len(tmpLen) > 6 {
+		firstByte++
+		firstByte = firstByte << 6
+		secondByte = uint8(tmpLen)
+		totalSize++ // secondbyte is optional
+		result = allocateBuffer(totalSize)
+		err = binary.Write(result, defaultByteOrder, firstByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+		err = binary.Write(result, defaultByteOrder, secondByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+	} else {
+		result = allocateBuffer(totalSize)
+		firstByte = firstByte << 1
+		firstByte += uint8(tmpLen)
+		err = binary.Write(result, defaultByteOrder, firstByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+	}
+
+	data, err := array.serialize()
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to serialize a bitarray")
+	}
+	err = binary.Write(result, defaultByteOrder, data)
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+	}
+
+	return result.Bytes(), nil
 }
 
 func (dbs *TriStateBitSet) Deserialize(data io.Reader) error {
+	var firstbyte uint8
+	err := binary.Read(data, defaultByteOrder, &firstbyte)
+	if err != nil {
+		return errors.Wrap(err, "[ Deserialize ] failed to read first byte")
+	}
+	compressed, hbitFlag, length := parseFirstByte(firstbyte)
+	if hbitFlag {
+		err = binary.Read(data, defaultByteOrder, &length)
+		if err != nil {
+			return errors.Wrap(err, "[ Deserialize ] failed to read second byte")
+		}
+	}
+	blockCount := uint64(math.Round(float64(length/8) + 0.5))
+	payload := make([]uint8, blockCount)
+	for i := 0; uint64(i) < blockCount; i++ {
+		err = binary.Read(data, defaultByteOrder, &payload[i])
+		if err != nil {
+			return errors.Wrap(err, "[ Deserialize ] failed to read first byte")
+		}
+	}
+	var array *bitArray
+	if compressed {
+		// TODO: uncompress payload
+	} else {
+		array, err = parseBitArray(payload, int(length))
+		if err != nil {
+			return nil
+		}
+	}
+	cells, err := dbs.parseCells(array)
+	if err != nil {
+		return err
+	}
+
+	dbs.cells = cells
+	dbs.CompressedSet = compressed
 	return nil
 }
 
+func (dbs *TriStateBitSet) parseCells(array *bitArray) ([]BitSetCell, error) {
+	cellSize := int(array.bitsSize / 2)
+	cells := make([]BitSetCell, cellSize)
+	for i := 0; i < cellSize; i++ {
+		id, err := dbs.mapper.IndexToRef(i)
+		if err != nil {
+			return nil, err
+		}
+		cells[i].NodeID = id
+		stateFirstBit, err := array.get(2 * i)
+		if err != nil {
+			return nil, err
+		}
+		stateSecondBit, err := array.get(2*i + 1)
+		if err != nil {
+			return nil, err
+		}
+		cells[i].State = TriState((stateFirstBit << 1) + stateSecondBit)
```

Comment:
Move lines 168-176 to separate func, e.g. `getTriState(array *bitArray, index int) TriState`. Same for setting `TriState`

Reply:
done